### PR TITLE
[RPS-421] CMS preview S3 proxy

### PR DESCRIPTION
**Pull request by wojtek-oledzki**
S3-hosted files can now be downloaded via CMS preview (using custom
pipeline and valve).

Also, S3ConnectorImpl is now unit tested. Logic generating S3 object
references has been extracted from S3ConnectorImpl into a standalone
class for testability.

### CODE REVIEW

**Comment by wojtek-oledzki** [comment]
@mchatzi can you take a look at 
https://github.com/NHS-digital-website/hippo/blob/387a4e8079f210d6e1751eefc9d39afe0dcce242/cms/src/main/java/uk/nhs/digital/externalstorage/resource/ImageDisplayPlugin.java#L97-L102

How can we get the "/_cmsinternal/s3endpoint/" path from sitemap?

File: site/src/test/java/uk/nhs/digital/test/util/RandomHelper.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+import java.util.stream.IntStream;
+
+import static java.util.Collections.shuffle;
+import static java.util.stream.Collectors.toList;
+
+public class RandomHelper {
```
Comment (ngta1) [inline site/src/test/java/uk/nhs/digital/test/util/RandomHelper.java]: Have a similar RandomHelper file in acceptance-test module.  CMS module could do with something similar for all the newRandomStrings. Could perhaps have 1 place for all these at some point going forward although not a priority for now though.

File: s3connector/src/main/java/uk/nhs/digital/externalstorage/s3/S3Connector.java [hunk 2]
```diff
@@ -15,5 +15,20 @@
 
     S3ObjectMetadata uploadFile(InputStream fileStream, String objectPath, String contentType);
 
-    long getFileSize(String objectPath);
+    /**
+     * <p>
+     * Retrieves a file identified by the parameter from S3.
+     * </p>
+     * <p>
+     * Aim to call this method once and make sure to close the stream returned returned by
```
Comment (ngta1) [inline s3connector/src/main/java/uk/nhs/digital/externalstorage/s3/S3Connector.java]: remove one of the "returned" in comment

File: site/src/test/java/uk/nhs/digital/common/valve/S3ConnectorValveTest.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        }
+
+        @Override
+        public int read(final byte[] buffer) throws IOException {
+            throwExceptinIfProgrammedTo();
+
+            return byteArrayInputStream.read(buffer);
+        }
+
+        void throwExceptinIfProgrammedTo() {
```
Comment (ngta1) [inline site/src/test/java/uk/nhs/digital/common/valve/S3ConnectorValveTest.java]: Exception

File: site/src/test/java/uk/nhs/digital/test/util/RandomHelper.java (hunk 1 above)
Comment (ziemsky) [reply site/src/test/java/uk/nhs/digital/test/util/RandomHelper.java]: I totally, agree, @ngta1.

I'll be raising a JIRA ticket to sort this out. I'm thinking of some shared 'test utils' or 'test resources' module.

**Pull request merged**